### PR TITLE
Add the plugin manager

### PR DESCRIPTION
**Pull request by pvdrz**

### CODE REVIEW

File: bee-node/Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,7 @@ homepage = "https://www.iota.org"
 bee-logger = { path = "../bee-common/bee-logger", features = [ "serde" ] }
 
 log = "0.4"
+notify = { version = "4.0" }
```
Comment (thibault-martinez) [inline bee-node/Cargo.toml]:
```suggestion
notify = "4.0"
```

File: bee-node/src/lib.rs [hunk 2]
```diff
@@ -9,3 +9,6 @@ pub mod banner;
 pub mod cli;
 pub mod config;
 pub mod constants;
+pub mod plugin;
+
+pub use banner::print_logo_and_version;
```
Comment (thibault-martinez) [inline bee-node/src/lib.rs]: Why ?

File: bee-node/src/plugin/manager.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+use std::{
+    collections::HashMap,
+    io,
+    path::{Path, PathBuf},
+    process::{Child, Command, Stdio},
+    sync::mpsc::{channel, Receiver},
+};
+
+use notify::{watcher, DebouncedEvent, RecommendedWatcher, RecursiveMode, Watcher};
```
Comment (thibault-martinez) [inline bee-node/src/plugin/manager.rs]: 3rd parties before std

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: bee-node/src/plugin/manager.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            }
+        }
+
+        Ok(())
+    }
+
+    fn load_plugin(&mut self, path: &Path) -> Result<(), io::Error> {
+        self.unload_plugin(path);
+
+        println!("loading plugin \"{}\"", path.display());
```
Comment (thibault-martinez) [inline bee-node/src/plugin/manager.rs]: Can you use the logger instead ?

File: bee-node/src/plugin/mod.rs [hunk 5]
```diff
@@ -0,0 +1,6 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! The bee plugin system.
```
Comment (thibault-martinez) [inline bee-node/src/plugin/mod.rs:4]:
```suggestion
//! The bee node plugin system.
```

File: bee-node/src/plugin/manager.rs [hunk 6]
```diff
@@ -0,0 +1,147 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! The bee plugin manager.
```
Comment (thibault-martinez) [inline bee-node/src/plugin/manager.rs:4]:
```suggestion
//! The bee node plugin manager.
```

File: bee-node/src/plugin/manager.rs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+/// The plugin manager.
+pub struct PluginManager {
+    plugins: HashMap<PathBuf, Child>,
+    rx: Receiver<DebouncedEvent>,
+    _watcher: RecommendedWatcher,
+}
+
+impl PluginManager {
+    /// Create a new plugin manager.
```
Comment (thibault-martinez) [inline bee-node/src/plugin/manager.rs:60]:
```suggestion
    /// Creates a new plugin manager.
```

File: bee-node/src/plugin/manager.rs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        for entry in entries {
+            let entry = entry?;
+            let path = entry.path().canonicalize()?;
+            self.load_plugin(&path)?;
+        }
+
+        Ok(())
+    }
+
+    /// Listen for changes in the plugins folder and loads every new plugin found in it.
```
Comment (thibault-martinez) [inline bee-node/src/plugin/manager.rs:86]:
```suggestion
    /// Listens for changes in the plugins folder and loads every new plugin found in it.
```

File: bee-node/src/plugin/manager.rs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+        while let Ok(event) = self.rx.recv() {
+            match event {
+                DebouncedEvent::Create(path)
+                | DebouncedEvent::Write(path)
+                | DebouncedEvent::Chmod(path)
+                | DebouncedEvent::Rename(_, path) => {
+                    let path = path.canonicalize()?;
+                    if let Err(err) = self.load_plugin(&path) {
+                        info!("plugin \"{}\" could not be loaded: {}", path.display(), err);
```
Comment (thibault-martinez) [inline bee-node/src/plugin/manager.rs:98]:
```suggestion
                        warn!("plugin \"{}\" could not be loaded: {}", path.display(), err);
```

File: bee-node/src/plugin/manager.rs [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+                    for (path, child) in self.plugins.drain() {
+                        Self::kill_plugin(&path, child);
+                    }
+
+                    self.scan()?;
+                }
+                DebouncedEvent::NoticeWrite(_)
+                | DebouncedEvent::NoticeRemove(_)
+                | DebouncedEvent::Error(_, _)
+                | DebouncedEvent::Remove(_) => (),
```
Comment (thibault-martinez) [inline bee-node/src/plugin/manager.rs:111]: Why not _ ?

File: bee-node/src/plugin/manager.rs [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+                DebouncedEvent::Create(path)
+                | DebouncedEvent::Write(path)
+                | DebouncedEvent::Chmod(path)
+                | DebouncedEvent::Rename(_, path) => {
+                    let path = path.canonicalize()?;
+                    if let Err(err) = self.load_plugin(&path) {
+                        info!("plugin \"{}\" could not be loaded: {}", path.display(), err);
+                    }
+                }
+                DebouncedEvent::Rescan => {
```
Comment (thibault-martinez) [inline bee-node/src/plugin/manager.rs:101]: When is that triggered ?

File: bee-node/src/plugin/manager.rs [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    fn load_plugin(&mut self, path: &Path) -> Result<(), io::Error> {
+        self.unload_plugin(path);
+
+        info!("loading plugin \"{}\"", path.display());
+        let child = Command::new(path)
+            .stdin(Stdio::null())
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .spawn()?;
+        info!("plugin \"{}\" loaded with PID {}", path.display(), child.id());
```
Comment (thibault-martinez) [inline bee-node/src/plugin/manager.rs:127]:
```suggestion
        debug!("plugin \"{}\" loaded with PID {}", path.display(), child.id());
```

File: bee-node/src/plugin/manager.rs [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+        assert!(self.plugins.insert(path.to_owned(), child).is_none());
```
Comment (thibault-martinez) [inline bee-node/src/plugin/manager.rs:129]: A little comment why we think this is never going to panic would be nice

File: bee-node/src/plugin/manager.rs [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    fn unload_plugin(&mut self, path: &Path) {
+        if let Some(child) = self.plugins.remove(path) {
+            Self::kill_plugin(path, child);
+        }
+    }
+
+    fn kill_plugin(path: &Path, mut child: Child) {
+        info!("killing plugin \"{}\"", path.display());
```
Comment (thibault-martinez) [inline bee-node/src/plugin/manager.rs:141]: I would put an unloading log in the other method rather than a killing load, for consistency

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Comment by pvdrz** [comment]
I think the design at https://github.com/iotaledger/bee/pull/668 is a bit better so I'm closing this PR in favor of the other one.

**Pull request closed without merging**